### ISSUE TITLE
`exported` rule does not warn of method comments not starting with the method name if the receiver is private

### ISSUE THREAD

**Issue by kostakhv** [mgechev/revive#552]
**Describe the bug**

`exported` rule does not warn of method comments not starting with the method name if the receiver is private. 
Probably this is the root cause- https://github.com/mgechev/revive/blob/98c374dcad22caddb52ea2a38394a4158a1de0c4/rule/exported.go#L64

**To Reproduce**
1. I updated revive `go get -u github.com/mgechev/revive`
2. I run it with the following flags & configuration file: default. 
3. Run revive on the following sample source code file, 

```go
package comments

type private struct {
}
type Public struct {
}

// this comment is wrong!
func (p *private) Method() {
}

// this comment is wrong!
func (p *Public) Method() {
}
```

```shell
# ./revive -version
Current revive version 1.0.9 commit 89f108c22aa9ccc72f190c8622578e52ee615361, built @2021-07-12T04:32:48Z by goreleaser.

# ./revive ./pkg/comments/...
pkg/comments/comments.go:5:6: exported type Public should have comment or be unexported
pkg/comments/comments.go:12:1: comment on exported method Public.Method should be of the form "Method ..."
```

**Expected behavior**
While `private` is not exported, therefore we do not expect any comment as this type. 
Its `Method` is exported and the wrong comment should be detected (as it is detected for `Public`'s method).

So basically I expected it to detect it as does GoLand for example, 

![image](https://user-images.githubusercontent.com/49071222/128343800-d7275d73-86f9-4868-985a-d3ff51b3900b.png)

**Desktop (please complete the following information):**
 - OS: Ubuntu 20.04.2
 - Version of Go: 1.16.6

**Comment by chavacava** [comment]
Hi @kostakhv, thanks for filling the issue. 

As you mention, the rule does not spot the "problem" because of the line you referenced. In fact these lines are equivalent to those in the `golint` code base (i.e. `revive` behaves as `golint`):

https://github.com/golang/lint/blob/6edffad5e6160f5949cdefc81710b2706fbcd4f6/lint.go#L859-L862

`golint` made the choice of not to lint comments of private identifiers (structs, functions, methods, ...)

In the case of the example you provided, even if `Method` of `private` is declared public (by upper-casing the first letter of its id) the method is actually private. Attaching a public method to a private struct is somewhat contradictory, and that is why the code of `golint` includes the above condition.

**Comment by kostakhv** [comment]
@chavacava thanks a lot for the detailed explanation. 

I think it's a pretty common practice to have a ctor return a private type, so that the caller can use it only to call public methods. 

```go
type object struct {
}

func NewObject(args ...string) *object {
	return &object{ /* init object with args */ }
}

func (o *object) Method() {
	// This method can be called from outside this package
}
```

https://stackoverflow.com/questions/37952354/golang-public-method-to-private-struct-is-this-have-any-use-case


Is there a chance we could add a config flag for this rule?
Something like,
```
privateTypeMethods = false // Check public methods of unexported types (default false).
```

**Comment by mgechev** [comment]
Adding a config flag and keeping the default aligned with golint sounds good to me.

**Comment by chavacava** [comment]
Hi @kostakhv, the branch [exported-config](https://github.com/mgechev/revive/tree/exported-config) adds the possibility of configuring the `exported` rule. 
There is a flag (_checkPrivateReceivers_) that enables checking of public methods of private types.

The configuration of the rule should be something like:

```toml
[rule.exported]
  arguments =["checkPrivateReceivers"]
```

Please let me know if it fits your needs.

**Comment by kostakhv** [comment]
@chavacava this looks awesome, appreciate it! 
Thanks a lot.
